<h1 align="center"> Homework 9 Programming Task (10 Points) </h1>
<h3 align="center"> IST 5520, Fall 2022 </h3>
<h3 align="center"> By:  </h3>
<h3 align="center"> Last Update: 10/30/2022 </h3>

The data file *Banks.csv* contains data on a sample of 20 banks. The “Financial Condition” column records the judgment of an expert on the financial condition of each bank. This outcome variable takes one of two possible values — 1 (weak) or 0 (strong) — according to the financial condition of the bank. The predictors are two ratios used in the financial analysis of banks: 

- **TotLns&Lses/Assets** is the ratio of total loans and leases to total assets;
- **TotExp/Assets** is the ratio of total expenses to total assets. The target is to use the two ratios for classifying the financial condition of a new bank.

## Step 1: Model Specification (3 points)

Use LaTeX syntax (for an introduction, refer to https://patrickwalls.github.io/mathematicalpython/jupyter/latex/) to write the estimated equation that associates the financial condition of a bank with its two predictors in three formats:

1)	The logit as a function of the predictors (1 point)

Your answer:  $logit(Y) = ln(\frac{Y}{1-Y}) = \beta_0 + \beta_1X_1 + \beta_2X_2$

2)	The odds as a function of the predictors (1 point) (This should be called odds ratio)

Your answer: $odds(Y=1) = \frac{Y}{1-Y} = e^{\beta_0 + \beta_1X_1 + \beta_2X_2}$

3)	The probability as a function of the predictors (1 point)

Your answer: $Pr(Y=1) = \frac{e^{\beta_0 + \beta_1X_1 + \beta_2X_2}}{1+e^{\beta_0 + \beta_1X_1 + \beta_2X_2}}$

## Step 2: Logistic Regression (5 points)

Implement a logistic regression model (on the entire dataset) that models the status of a bank as a function of the two financial measures provided. Interpret the estimates of coefficients.

Note: Feel free to add more code lines below to implement the logic.

#### Import data

In [71]:
import pandas as pd
import numpy as np

In [72]:
df = pd.read_csv('banks.csv')

df.head()

Obs  Financial Condition  TotCap/Assets  TotExp/Assets  TotLns&Lses/Assets
0    1                    1            9.7           0.12                0.65
1    2                    1            1.0           0.11                0.62
2    3                    1            6.9           0.09                1.02
3    4                    1            5.8           0.10                0.67
4    5                    1            4.3           0.11                0.69

#### Set up logistic regression

In [73]:
import statsmodels.formula.api as sm

In [74]:
#rename dataframe columns to work w/ logistic regression, set up formula for logistic regression
df = df.rename(columns = {df.columns[1]: 'Financial_Condition', df.columns[3]: 'TotExp_Assets',\
                     df.columns[4]: 'TotLns_Lses_Assets'})
    
print(df.columns)

formula = 'Financial_Condition ~ TotExp_Assets + TotLns_Lses_Assets'

Index(['Obs', 'Financial_Condition', 'TotCap/Assets', 'TotExp_Assets',
       'TotLns_Lses_Assets'],
      dtype='object')


In [75]:
#fit logistic regression onto data
log_reg = sm.logit(formula, data=df).fit()

Optimization terminated successfully.
         Current function value: 0.328688
         Iterations 8


#### print summary statistics

In [76]:
display(log_reg.summary())

<class 'statsmodels.iolib.summary.Summary'>
"""
                            Logit Regression Results                           
===============================================================================
Dep. Variable:     Financial_Condition   No. Observations:                   20
Model:                           Logit   Df Residuals:                       17
Method:                            MLE   Df Model:                            2
Date:                 Sun, 06 Nov 2022   Pseudo R-squ.:                  0.5258
Time:                         18:51:33   Log-Likelihood:                -6.5738
converged:                        True   LL-Null:                       -13.863
Covariance Type:             nonrobust   LLR p-value:                 0.0006829
======================================================================================
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
Intercept            -14.7210      6.675     -2.205      0.027     -27.804      -1.638
TotExp_Assets         89.8339     47.781      1.880      0.060      -3.815     183.483
TotLns_Lses_Assets     8.3713      5.779      1.449      0.147      -2.955      19.698
======================================================================================
"""

#### TotLns_Lses_Assets:

As the ratio of loans and leases to assets increases by one,
the likelihood of a weak financial condition increases by $e^{8.3713}$ = 4321 times.

#### TotExp_Assets:

As the ratio of expenses to assets increases by one,
the likelihood a weak financial condition increases by $e^{89.8339}$ = 1.0336 * $10^{39}$ times

Calculate marginal effects and interpret the estimates of marginal effects.

In [77]:
log_reg.get_margeff().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
        Logit Marginal Effects        
======================================
Dep. Variable:     Financial_Condition
Method:                           dydx
At:                            overall
======================================================================================
                        dy/dx    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
TotExp_Assets          9.2747      3.028      3.063      0.002       3.340      15.210
TotLns_Lses_Assets     0.8643      0.467      1.850      0.064      -0.051       1.780
======================================================================================
"""

#### TotExp_Assets

As the ratio of Expenses to Assets increases by one, the likelihood of a weak financial condition increases by 927.47%

#### TotLns_Lses_Assets

As the ratio of Loans and Leases to Assets increases by one, the likelihood a weak financial condition increases by 86.43%

## Step 3. Predict New Data (2 points)

Consider a new bank whose total loans and leases/assets ratio = 0.6 and total expenses/assets ratio = 0.11. From your logistic regression model, estimate the following two quantities for this bank: 

- the probability of being financially weak;
- the classification of the bank (use cutoff = 0.5).


In [78]:
#predict test data
data = {'TotExp_Assets': [.11], 'TotLns_Lses_Assets': [.6]}

scenario = pd.DataFrame(data)

scen_pred = log_reg.predict(scenario)

In [79]:
#Label prediction
scenario['Prediction'] = ['Weak' if x >= .5 else 'Strong' for x in scen_pred]

#print result
scenario

TotExp_Assets  TotLns_Lses_Assets Prediction
0           0.11                 0.6       Weak